File: go/adapter_pattern.go
package main

import "fmt"

// Existing implementation
type AlgoliaProducts struct {
	productVariants []string
}

// Define interface to support both existing and new structures
type ProductsInterface interface {
	getProductVariants()
}

// Implement interface for AlgoliaProducts
func (a *AlgoliaProducts) getProductVariants() {
	fmt.Println("Algolia StoreProducts")
}

// New schema
type AtlasProducts struct {
	productVariants []string
}

// Atlas-specific method to represent the data
func (a *AtlasProducts) getAtlasProductVariants() {
	fmt.Println("Atlas StoreProducts")
}

// Adapter to make AtlasProducts compatible with ProductsInterface
type AtlasProductsAdapter struct {
	atlasProducts *AtlasProducts
}

// Implement ProductsInterface for AtlasProductsAdapter
func (a *AtlasProductsAdapter) getProductVariants() {
	a.atlasProducts.getAtlasProductVariants()
}

func main() {
	algolia := &AlgoliaProducts{}
	algolia.getProductVariants()

	atlas := &AtlasProducts{}
	atlasAdapter := &AtlasProductsAdapter{
		atlasProducts: atlas,
	}
	atlasAdapter.getProductVariants()
}

File: java/adapter_pattern/AdapterPattern.java
interface ProductsInterface {
    void getProductVariants();
}

class AlgoliaProducts implements ProductsInterface {
    @Override
    public void getProductVariants() {
        System.out.println("Algolia StoreProducts");
    }
}

class AtlasProducts {
    public void getAtlasProductVariants() {
        System.out.println("Atlas StoreProducts");
    }
}

class AtlasProductsAdapter implements ProductsInterface {
    private final AtlasProducts atlasProducts;

    public AtlasProductsAdapter(AtlasProducts atlasProducts) {
        this.atlasProducts = atlasProducts;
    }

    @Override
    public void getProductVariants() {
        atlasProducts.getAtlasProductVariants();
    }
}

public class AdapterPattern {
    public static void main(String[] args) {
        ProductsInterface algolia = new AlgoliaProducts();
        algolia.getProductVariants();

        AtlasProducts atlas = new AtlasProducts();
        ProductsInterface atlasAdapter = new AtlasProductsAdapter(atlas);
        atlasAdapter.getProductVariants();
    }
}
